README
TinyMCE Embed Plug-in

Embed Flash, Iframes etc. into TinyMCE

Tested with: TinyMCE 3.2.7, TinyMCE 3.3.8
Firefox 3.6.6, Opera 10, Safari 5, Chrome 5, IE 8, IE 7, IE 6

Installation
============
* Make sure the media plugin is added to the plug-ins configuration.
* Extract the zip to the TinyMCE plug-ins folder.
* Add embed to the plug-ins configuration.
* Add embed to the theme_advanced_buttons_n configuration.

Optional Configuration
======================
Add these to the TinyMCE configuration object:

embed_iframe_innerhtml_fallback : <String> - If the iframe tags contains no content, this string will be added. If not set "This browser does not support the iframe element." will be added.

changelog.txt
Version 1.1
    - Dialog insert button value is changed to update if an embedded element is selected

Version 1.0
    - Initial release

tests/test_embed.js
var editor;
var embed;

QUnit.config.autostart = false;

test("Insert Google Maps", function() {
    expect(2);

    editor.setContent('<iframe width="640" height="480" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="http://maps.google.com/?ie=UTF8&amp;t=h&amp;ll=63.548552,27.685547&amp;spn=19.042129,56.25&amp;z=4&amp;output=embed"></iframe><br /><small><a href="http://maps.google.com/?ie=UTF8&amp;t=h&amp;ll=63.548552,27.685547&amp;spn=19.042129,56.25&amp;z=4&amp;source=embed" style="color:#0000FF;text-align:left">View Larger Map</a></small>');
    equals(editor.getContent(), '<p><iframe frameborder="0" height="480" marginheight="0" marginwidth="0" scrolling="no" src="http://maps.google.com/?ie=UTF8&amp;t=h&amp;ll=63.548552,27.685547&amp;spn=19.042129,56.25&amp;z=4&amp;output=embed" width="640">This browser does not support the iframe element.</iframe><br /><small><a style="color: #0000ff; text-align: left;" href="http://maps.google.com/?ie=UTF8&amp;t=h&amp;ll=63.548552,27.685547&amp;spn=19.042129,56.25&amp;z=4&amp;source=embed">View Larger Map</a></small></p>');

    editor.setContent('<iframe frameborder="0" height="350" marginheight="0" marginwidth="0" scrolling="no" src="http://maps.google.com/?ie=UTF8&amp;ll=25.085599,-95.712891&amp;spn=80.038802,191.513672&amp;t=h&amp;z=4&amp;output=embed" width="425"><p>Custom innerHTML</p></iframe>');
    equals(editor.getContent(), '<p><iframe frameborder="0" height="350" marginheight="0" marginwidth="0" scrolling="no" src="http://maps.google.com/?ie=UTF8&amp;ll=25.085599,-95.712891&amp;spn=80.038802,191.513672&amp;t=h&amp;z=4&amp;output=embed" width="425"><p>Custom innerHTML</p></iframe></p>');
});

test("Insert Bing Maps", function() {
    expect(2);

    editor.setContent('<div id="mapviewer"><iframe id="map" Name="mapFrame" scrolling="no" width="500" height="400" frameborder="0" src="http://www.bing.com/maps/embed/?lvl=3&amp;cp=57.564380295549434~31.580111503156132&amp;sty=r&amp;draggable=true&amp;v=2&amp;emid=2d014405-e05c-1bdb-cabb-9fbe8a610a3b&amp;w=500&amp;h=400"></iframe></div>');
    equals(editor.getContent(), '<div id="mapviewer"><iframe frameborder="0" height="400" id="map" name="mapFrame" scrolling="no" src="http://www.bing.com/maps/embed/?lvl=3&amp;cp=57.564380295549434~31.580111503156132&amp;sty=r&amp;draggable=true&amp;v=2&amp;emid=2d014405-e05c-1bdb-cabb-9fbe8a610a3b&amp;w=500&amp;h=400" width="500">This browser does not support the iframe element.</iframe></div>');

    editor.setContent('<iframe id="map" Name="mapFrame" scrolling="no" width="500" height="400" frameborder="0" src="http://www.bing.com/maps/embed/?lvl=3&amp;cp=57.564380295549434~31.580111503156132&amp;sty=r&amp;draggable=true&amp;v=2&amp;emid=2d014405-e05c-1bdb-cabb-9fbe8a610a3b&amp;w=500&amp;h=400"><p>Custom innerHTML</p></iframe>');
    equals(editor.getContent(), '<p><iframe frameborder="0" height="400" id="map" name="mapFrame" scrolling="no" src="http://www.bing.com/maps/embed/?lvl=3&amp;cp=57.564380295549434~31.580111503156132&amp;sty=r&amp;draggable=true&amp;v=2&amp;emid=2d014405-e05c-1bdb-cabb-9fbe8a610a3b&amp;w=500&amp;h=400" width="500"><p>Custom innerHTML</p></iframe></p>');
});

test("Insert Google Docs", function() {
    expect(1);

    editor.setContent('<iframe scrolling="no" width="500" height="400" frameborder="0" src="http://spreadsheets.google.com/pub?key=0AuQGDiRNeISsdHk4NlRHX1NaWXU2N01ZYmYxWTc4MWc&hl=en&single=true&gid=0&output=html"><p>Custom innerHTML</p></iframe>');
    equals(editor.getContent(), '<p><iframe frameborder="0" height="400" scrolling="no" src="http://spreadsheets.google.com/pub?key=0AuQGDiRNeISsdHk4NlRHX1NaWXU2N01ZYmYxWTc4MWc&amp;hl=en&amp;single=true&amp;gid=0&amp;output=html" width="500"><p>Custom innerHTML</p></iframe></p>');
});

test("Insert YouTube Player (iframe version)", function() {
    expect(1);

    editor.setContent('<iframe class="youtube-player" type="text/html" width="640" height="385" src="http://www.youtube.com/embed/EQt54x6KmK8" frameborder="0">\n</iframe>');
    equals(editor.getContent(), '<p><iframe class="youtube-player" frameborder="0" height="385" src="http://www.youtube.com/embed/EQt54x6KmK8" width="640" type="text/html">This browser does not support the iframe element.</iframe></p>');
});

test("Insert YouTube Player", function() {
    expect(1);

    editor.setContent('<object width="640" height="385"><param name="movie" value="http://www.youtube.com/v/EQt54x6KmK8&amp;hl=en_US&amp;fs=1"></param><param name="allowFullScreen" value="true"></param><param name="allowscriptaccess" value="always"></param><embed src="http://www.youtube.com/v/EQt54x6KmK8&amp;hl=en_US&amp;fs=1" type="application/x-shockwave-flash" allowscriptaccess="always" allowfullscreen="true" width="640" height="385"></embed></object>');
    equals(editor.getContent(), '<p><object width="640" height="385" data="http://www.youtube.com/v/EQt54x6KmK8&amp;hl=en_US&amp;fs=1" type="application/x-shockwave-flash"><param name="allowFullScreen" value="true" /><param name="allowscriptaccess" value="always" /><param name="src" value="http://www.youtube.com/v/EQt54x6KmK8&amp;hl=en_US&amp;fs=1" /><param name="allowfullscreen" value="true" /></object></p>');
});

test("Insert Vimeo Player", function() {
    expect(1);

    editor.setContent('<object width="400" height="155"><param name="allowfullscreen" value="true" /><param name="allowscriptaccess" value="always" /><param name="movie" value="http://vimeo.com/moogaloop.swf?clip_id=13616392&amp;server=vimeo.com&amp;show_title=1&amp;show_byline=1&amp;show_portrait=0&amp;color=&amp;fullscreen=1" /><embed src="http://vimeo.com/moogaloop.swf?clip_id=13616392&amp;server=vimeo.com&amp;show_title=1&amp;show_byline=1&amp;show_portrait=0&amp;color=&amp;fullscreen=1" type="application/x-shockwave-flash" allowfullscreen="true" allowscriptaccess="always" width="400" height="155"></embed></object><p><a href="http://vimeo.com/13616392">2 nights in Mallory Square</a> from <a href="http://vimeo.com/philipbloom">Philip Bloom</a> on <a href="http://vimeo.com">Vimeo</a>.</p>');
    equals(editor.getContent(), '<p><object width="400" height="155" data="http://vimeo.com/moogaloop.swf?clip_id=13616392&amp;server=vimeo.com&amp;show_title=1&amp;show_byline=1&amp;show_portrait=0&amp;color=&amp;fullscreen=1" type="application/x-shockwave-flash"><param name="allowfullscreen" value="true" /><param name="allowscriptaccess" value="always" /><param name="src" value="http://vimeo.com/moogaloop.swf?clip_id=13616392&amp;server=vimeo.com&amp;show_title=1&amp;show_byline=1&amp;show_portrait=0&amp;color=&amp;fullscreen=1" /></object></p><p><a href="http://vimeo.com/13616392">2 nights in Mallory Square</a> from <a href="http://vimeo.com/philipbloom">Philip Bloom</a> on <a href="http://vimeo.com">Vimeo</a>.</p>');
});

test("Iframe output is valid XHTML Transitional", function() {
    expect(1);

    editor.setContent('<iframe invalid-attribute="foo" class="foo-bar" longdesc="This is a Google Spreadsheet" name="g-sh" marginwidth="9" marginheight="0" scrolling="no" align="top" id="g-sh" style="margin:10px ;border:1px solid black" title="Google Docs" width="500" height="400" frameborder="0" src="http://spreadsheets.google.com/pub?key=0AuQGDiRNeISsdHk4NlRHX1NaWXU2N01ZYmYxWTc4MWc&hl=en&single=true&gid=0&output=html" invalid-attribute2="bar">Custom innerHTML</iframe>');
    equals(editor.getContent(), '<p><iframe align="top" class="foo-bar" frameborder="0" height="400" id="g-sh" longdesc="This is a Google Spreadsheet" marginheight="0" marginwidth="9" name="g-sh" scrolling="no" src="http://spreadsheets.google.com/pub?key=0AuQGDiRNeISsdHk4NlRHX1NaWXU2N01ZYmYxWTc4MWc&amp;hl=en&amp;single=true&amp;gid=0&amp;output=html" style="margin: 10px; border: 1px solid black;" title="Google Docs" width="500">Custom innerHTML</iframe></p>');
});





editor_plugin.js
(function(){tinymce.PluginManager.requireLangPack("embed");var a=/^(mceItemIframe|mceItemFlash|mceItemShockWave|mceItemWindowsMedia|mceItemQuickTime|mceItemRealMedia)$/;tinymce.create("tinymce.plugins.embed",{init:function(c,d){var e=this;function b(f){return a.test(f.className)}e.url=d;e.editor=c;e.embed_iframe_innerhtml_fallback=(c.settings.embed_iframe_innerhtml_fallback)?c.settings.embed_iframe_innerhtml_fallback:"This browser does not support the iframe element.";c.onPreInit.add(function(){c.serializer.addRules("iframe[_iframe_innerhtml|align<bottom?left?middle?right?top|class|frameborder|height|id|longdesc|marginheight|marginwidth|name|scrolling<auto?no?yes|src|style|title|width|type]")});c.onInit.add(function(){if(c.settings.content_css!==false){c.dom.loadCSS(d+"/css/embed.css")}});c.addCommand("enonicEmbed",function(){c.windowManager.open({file:d+"/window.html",width:620+c.getLang("embed.delta_width",0),height:540+c.getLang("embed.delta_height",0),inline:1},{plugin_url:d})});c.addButton("embed",{title:"embed.desc",cmd:"enonicEmbed",image:d+"/img/embed.gif"});c.onNodeChange.add(function(g,f,h){f.setActive("embed",b(h))});c.onBeforeSetContent.add(function(f,g){g.content=e.iframesToSpans(g.content)});c.onSetContent.add(function(f,g){e.spansToImages(g.node)});c.onPreProcess.add(function(f,g){if(g.set){g.content=e.iframesToSpans(g.content);e.spansToImages(g.node)}if(g.get){e.imagesToIframes(g)}});c.onPostProcess.add(function(f,g){if(g.get){g.content=g.content.replace(/(<iframe.+?)_iframe_innerhtml="(.+?)"(.+?)(<\/iframe>)/gi,function(){var h=e.editor.dom.decode(arguments[2]).replace(/&lt;/gm,"<").replace(/&gt;/gm,">");return arguments[1]+tinymce.trim(arguments[3])+h+arguments[4]})}})},getInfo:function(){return{longname:"Embed Plug-in",author:"tan@enonic.com",authorurl:"http://www.enonic.com",infourl:"http://www.enonic.com",version:"1.1"}},iframesToSpans:function(c){var b=this;return c.replace(/<iframe\s*(.*?)>(|[\s\S]+?)<\/iframe>/gim,function(){var d="<span ";d+=arguments[1];d+=' _class="mceItemIframe">';d+=b.editor.dom.encode(arguments[2]);d+="</span>";return d})},spansToImages:function(e){var c=this,d=c.editor,f=d.dom,g;var b=f.select('span[_class="mceItemIframe"]',e);tinymce.each(b,function(h){g=c.createImagePlaceHolder(h);f.replace(g,h)})},imagesToIframes:function(g){var c=this,e=c.editor,f=e.dom,d;var b=f.select('img[class="mceItemIframe"]',g.node);tinymce.each(b,function(h){d=c.createIframeElement(h);f.replace(d,h)})},createImagePlaceHolder:function(i){var k=this,g=k.editor,f=g.dom,d,h,b,j,c,e;b=f.getAttrib(i,"width");j=f.getAttrib(i,"height");h=k._serializeIframeAttributes(i);e=i.innerHTML.replace(/^\s+|\s+$/g,"");e=e!==""?e:k.embed_iframe_innerhtml_fallback;e=e.replace(/<(.+?)>/gim,function(){return"<"+arguments[1].toLowerCase()+">"});h+=',"innerhtml":"'+e+'"';d=f.create("img");f.setAttrib(d,"src",k.url+"/img/trans.gif");f.setAttrib(d,"title",h);f.setAttrib(d,"width",b);f.setAttrib(d,"height",j);f.addClass(d,"mceItemIframe");f.setStyle(d,"width",b);f.setAttrib(d,"style",0);return d},createIframeElement:function(c){var j=this,g=j.editor,d=g.dom,h="",e,b,i;var f=j._parseImagePlaceHolderTitle(c);if("innerhtml" in f){h=f.innerhtml;delete f.innerhtml}b=d.getAttrib(c,"width");i=d.getAttrib(c,"height");e=d.create("iframe",f);d.setAttrib(e,"width",b);d.setAttrib(e,"height",i);d.setAttrib(e,"_iframe_innerhtml",h);return e},_parseImagePlaceHolderTitle:function(e){var c=this,d=c.editor,b=d.dom.getAttrib(e,"title");return tinymce.util.JSON.parse("{"+b+"}")},_serializeIframeAttributes:function(d){var j=this,h=j.editor,c=h.dom,g,f,k,e={};var b=["src","width","height","longdesc","name","frameborder","marginwidth","marginheight","scrolling","align","id","class","style","title","type"];for(var i in b){f=b[i];k=c.getAttrib(d,f);g=k!=="";if(g){e[f]=k}}return tinymce.util.JSON.serialize(e).replace(/[{}]/g,"")}});tinymce.PluginManager.add("embed",tinymce.plugins.embed)})();

js/embed_dialog.js
var g_editor;
var g_textarea, g_previewContainer;
var g_update = false;

var pathToCodeMirror = '../../../../examples/CodeMirror-0.8/';

function init()
{
    tinyMCEPopup.restoreSelection();

    tinyMCEPopup.resizeToInnerSize();

    g_editor = tinyMCEPopup.editor;
    g_textarea = document.getElementById('source');
    g_previewContainer = document.getElementById('preview');

    var selectedNode = g_editor.selection.getNode();
    var htmlFragment, iframeInnerHTML;

    var isPlaceholderElem = /^(mceItemIframe|mceItemFlash|mceItemShockWave|mceItemWindowsMedia|mceItemQuickTime|mceItemRealMedia)$/.test(selectedNode.className);

    if ( !isPlaceholderElem )
    {
        return;
    }

    if ( isPlaceholderElem )
    {
        g_update = true;

        document.getElementById('insert').value = g_editor.getLang('embed.btn_update');        
    }

    htmlFragment = transformPlaceholderElem( selectedNode );

    iframeInnerHTML = g_editor.dom.getAttrib(htmlFragment, '_iframe_innerhtml');
    iframeInnerHTML = g_editor.dom.decode(iframeInnerHTML);

    g_editor.dom.setAttrib(htmlFragment, '_iframe_innerhtml', '');

    htmlFragment = g_editor.serializer.serialize(htmlFragment);

    htmlFragment = htmlFragment.replace(/(<iframe\s.+?>)(|.+?)(<\/iframe>)/gi, '$1'+iframeInnerHTML+'$3');

    g_textarea.value = htmlFragment;

    /*
    var codemirror = CodeMirror.fromTextArea('source', {
        parserfile: ["parsexml.js", "parsecss.js", "tokenizejavascript.js", "parsejavascript.js", "parsehtmlmixed.js"],
        stylesheet: ["" + pathToCodeMirror + "css/xmlcolors.css", ""+pathToCodeMirror+"css/jscolors.css", ""+pathToCodeMirror+"css/csscolors.css"],
        path: pathToCodeMirror + "js/",
        tabMode: "indent",
        reindentOnLoad: true
    });
    */


    // Set time out since it seems Safari not always will update the preview on init.
    setTimeout(function() {
        updatePreview();
    }, 30);

}

function insertSource()
{
    var source = g_textarea.value;

    if ( window.tinyMCE )
    {
        if ( isSourceValid() )
        {
            if ( g_update )
            {
                tinyMCEPopup.restoreSelection();

                var dom = g_editor.dom;
                var selection = g_editor.selection;
                var bogus = g_editor.getDoc().createElement( 'br' );
                bogus.id = '___temp1';

                dom.replace( bogus, selection.getNode() );

                selection.select( bogus );
            }

            tinyMCE.execCommand('mceInsertRawHTML',false, source);

            tinyMCEPopup.editor.execCommand( 'mceRepaint' );
            tinyMCEPopup.close();
        }
        else
        {
            showError();
        }
    }
    return true;
}

function updatePreview()
{
    var previewIframe = g_previewContainer;

    if (isSourceValid())
    {
        previewIframe = (previewIframe.contentWindow) ? previewIframe.contentWindow : (previewIframe.contentDocument.document) ? previewIframe.contentDocument.document : previewIframe.contentDocument;
        previewIframe.document.open();
        previewIframe.document.write(g_textarea.value);
        previewIframe.document.close();
    }
    else
    {
        showError();
    }
}

function transformPlaceholderElem( node )
{
    if ( node.className == 'mceItemIframe' )
    {
        return tinyMCEPopup.editor.plugins.embed.createIframeElement( node );
    }

    var ci, cb, mt;

    var media = tinyMCEPopup.editor.plugins.media;

    switch ( node.className )
    {
        case 'mceItemFlash':
            ci = 'd27cdb6e-ae6d-11cf-96b8-444553540000';
            cb = 'http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=6,0,40,0';
            mt = 'application/x-shockwave-flash';
            break;

        case 'mceItemShockWave':
            ci = '166b1bca-3f9c-11cf-8075-444553540000';
            cb = 'http://download.macromedia.com/pub/shockwave/cabs/director/sw.cab#version=8,5,1,0';
            mt = 'application/x-director';
            break;

        case 'mceItemWindowsMedia':
            ci = ed.getParam( 'media_wmp6_compatible' ) ? '05589fa1-c356-11ce-bf01-00aa0055595a' : '6bf52a52-394a-11d3-b153-00c04f79faa6';
            cb = 'http://activex.microsoft.com/activex/controls/mplayer/en/nsmp2inf.cab#Version=5,1,52,701';
            mt = 'application/x-mplayer2';
            break;

        case 'mceItemQuickTime':
            ci = '02bf25d5-8c17-4b23-bc80-d3488abddc6b';
            cb = 'http://www.apple.com/qtactivex/qtplugin.cab#version=6,0,2,0';
            mt = 'video/quicktime';
            break;

        case 'mceItemRealMedia':
            ci = 'cfcdaa03-8be4-11cf-b84b-0020afbbccfa';
            cb = 'http://download.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=6,0,40,0';
            mt = 'audio/x-pn-realaudio-plugin';
            break;
    }

    if ( ci )
    {
        return media._buildObj( { classid : ci, codebase : cb, type : mt }, node );
    }

    return '';
}

function showError()
{
    alert('Source contains script. Please remove');
}

function isSourceValid()
{
    var valid = true;
    var source = g_textarea.value;

    if ( source.indexOf( '<script' ) > -1 )
    {
        valid = false;
    }

    return valid;
}

editor_plugin_src.js
/**
 * TinyMCE Embed plugin
 *
 * A plugin that let's users paste embeded code such as YouTube videos, Google Maps into the document.
 * NOTE: This plugin requires the TinyMCE Media plugin.
 *
 * Icon: Mark James, http://www.famfamfam.com/lab/icons/silk/ (CC 2.5)
 *
 */

(function()
{
    tinymce.PluginManager.requireLangPack( 'embed' );

    var imagePlaceHolderPattern = /^(mceItemIframe|mceItemFlash|mceItemShockWave|mceItemWindowsMedia|mceItemQuickTime|mceItemRealMedia)$/;

    tinymce.create( 'tinymce.plugins.embed', {
        init : function( ed, url )
        {
            var t = this;

            function isImagePlaceHolder( n )
            {
                return imagePlaceHolderPattern.test(n.className);
            }

            t.url = url;
            t.editor = ed;
            t.embed_iframe_innerhtml_fallback = ( ed.settings.embed_iframe_innerhtml_fallback ) ? ed.settings.embed_iframe_innerhtml_fallback : 'This browser does not support the iframe element.';

            ed.onPreInit.add(function() {
                ed.serializer.addRules('iframe[_iframe_innerhtml|align<bottom?left?middle?right?top|class|frameborder|height|id|longdesc|marginheight|marginwidth|name|scrolling<auto?no?yes|src|style|title|width|type]');
            });

            ed.onInit.add( function()
            {
                if ( ed.settings.content_css !== false )
                {
                    ed.dom.loadCSS( url + "/css/embed.css" );
                }
            } );

            ed.addCommand( 'enonicEmbed', function()
            {
                ed.windowManager.open( {
                    file : url + '/window.html',
                    width : 620 + ed.getLang( 'embed.delta_width', 0 ),
                    height : 540 + ed.getLang( 'embed.delta_height', 0 ),
                    inline : 1
                }, {
                    plugin_url : url
                } );
            } );

            ed.addButton( 'embed', {
                title : 'embed.desc',
                cmd : 'enonicEmbed',
                image : url + '/img/embed.gif'
            } );

            ed.onNodeChange.add( function( ed, cm, n )
            {
                cm.setActive( 'embed', isImagePlaceHolder( n ) );
            } );

            ed.onBeforeSetContent.add( function( ed, o )
            {
                o.content = t.iframesToSpans( o.content );
            });

            ed.onSetContent.add( function( ed, o )
            {
                t.spansToImages( o.node );
            } );

            ed.onPreProcess.add( function( ed, o )
            {
                if ( o.set )
                {
                    o.content = t.iframesToSpans( o.content );
                    t.spansToImages( o.node );
                }

                if ( o.get )
                {
                    t.imagesToIframes( o );
                }
            } );

            ed.onPostProcess.add( function( ed, o )
            {
                if ( o.get )
                {
                    o.content = o.content.replace(/(<iframe.+?)_iframe_innerhtml="(.+?)"(.+?)(<\/iframe>)/gi, function() {
                        var innerHTML = t.editor.dom.decode(arguments[2]).replace(/&lt;/gm, '<').replace(/&gt;/gm, '>');

                        return arguments[1] + tinymce.trim(arguments[3]) + innerHTML + arguments[4];
                    });
                }
            } );
        },


        getInfo : function()
        {
            return {
                longname  : 'Embed Plug-in',
                author    : 'tan@enonic.com',
                authorurl : 'http://www.enonic.com',
                infourl   : 'http://www.enonic.com',
                version   : "1.1"
            };
        },

        
        iframesToSpans : function( content )
        {
            var t = this;
            
            return content.replace(/<iframe\s*(.*?)>(|[\s\S]+?)<\/iframe>/gim, function() {
                var replacement = '<span ';

                replacement += arguments[1];
                replacement += ' _class="mceItemIframe">';
                replacement += t.editor.dom.encode(arguments[2]);
                replacement += '</span>';

                return replacement;
            });
        },


        spansToImages: function( node )
        {
            var t = this,
                    editor = t.editor,
                    dom = editor.dom,
                    imagePlaceHolder;

            var spans = dom.select( 'span[_class="mceItemIframe"]', node );

            tinymce.each( spans, function( span )
            {
                imagePlaceHolder = t.createImagePlaceHolder( span );

                dom.replace( imagePlaceHolder, span );
            } );
        },


        imagesToIframes : function( o )
        {
            var t = this,
                    editor = t.editor,
                    dom = editor.dom,
                    iframe;

            var imagePlaceHolders = dom.select( 'img[class="mceItemIframe"]', o.node );

            tinymce.each( imagePlaceHolders, function( img )
            {
                iframe = t.createIframeElement( img );
                dom.replace( iframe, img );
            } );
        },


        createImagePlaceHolder : function( span )
        {
            var t = this,
                    editor = t.editor,
                    dom = editor.dom,
                    image,
                    title,
                    width,
                    height,
                    iframeInnerHTMLAttrib,
                    iframeInnerHTML;

            width = dom.getAttrib( span, 'width' );
            height = dom.getAttrib( span, 'height' );
            title = t._serializeIframeAttributes( span );

            iframeInnerHTML = span.innerHTML.replace(/^\s+|\s+$/g, '');

            iframeInnerHTML = iframeInnerHTML !== '' ? iframeInnerHTML : t.embed_iframe_innerhtml_fallback;

            // Lowercase tags for IE.
            iframeInnerHTML = iframeInnerHTML.replace(/<(.+?)>/gim, function () {
                return '<'+arguments[1].toLowerCase()+'>';
            });

            title += ',"innerhtml":"'+iframeInnerHTML+'"';

            image = dom.create( 'img' );

            dom.setAttrib( image, 'src', t.url + '/img/trans.gif' );
            dom.setAttrib( image, 'title', title );
            dom.setAttrib( image, 'width', width );
            dom.setAttrib( image, 'height', height );
            dom.addClass( image, 'mceItemIframe' );

            // For some reason IE will not set the img.width attribute if the image is not fully loaded(onload).
            dom.setStyle( image, 'width', width );
            dom.setAttrib( image, 'style', 0 );

            return image;
        },

        
        createIframeElement : function( imagePlaceHolder )
        {
            var t = this,
                    editor = t.editor,
                    dom = editor.dom,
                    innerHTML = '',
                    iframe,
                    width,
                    height;

            var attribsForIframe = t._parseImagePlaceHolderTitle( imagePlaceHolder );
            
            if ( 'innerhtml' in attribsForIframe )
            {
                innerHTML = attribsForIframe.innerhtml;
                delete attribsForIframe.innerhtml;
            }

            width = dom.getAttrib( imagePlaceHolder, 'width' );
            height = dom.getAttrib( imagePlaceHolder, 'height' );
            iframe = dom.create( 'iframe', attribsForIframe );

            dom.setAttrib( iframe, 'width', width );
            dom.setAttrib( iframe, 'height', height );
            dom.setAttrib( iframe, '_iframe_innerhtml', innerHTML );

            return iframe;
        },


        _parseImagePlaceHolderTitle : function( imagePlaceHolder )
        {
            var t = this,
                    editor = t.editor,
                    placeHolderTitle = editor.dom.getAttrib( imagePlaceHolder, 'title' );

            return tinymce.util.JSON.parse('{' + placeHolderTitle + '}');
        },


        _serializeIframeAttributes : function( iframe )
        {
            var t = this,
                    editor = t.editor,
                    dom = editor.dom,
                    iframeHasValidAttrib,
                    validAttrib,
                    iframeAttrib,
                    attribsForPlaceHolder = {};

            var xhtml1TransitionalAttribs = ['src', 'width', 'height', 'longdesc', 'name', 'frameborder', 'marginwidth', 'marginheight', 'scrolling', 'align', 'id', 'class', 'style', 'title', 'type'];

            for ( var key in xhtml1TransitionalAttribs )
            {
                validAttrib = xhtml1TransitionalAttribs[key];
                iframeAttrib = dom.getAttrib( iframe, validAttrib );

                iframeHasValidAttrib = iframeAttrib !== '';

                if ( iframeHasValidAttrib )
                {
                    attribsForPlaceHolder[validAttrib] = iframeAttrib;
                }
            }

            return tinymce.util.JSON.serialize(attribsForPlaceHolder).replace(/[{}]/g, '');
        }
    } );

    tinymce.PluginManager.add( 'embed', tinymce.plugins.embed );
})();
